op-e2e/actions/op_program_test.go

```diff
@@ -0,0 +1,270 @@
+	l2ClCfg := sources.L2ClientDefaultConfig(env.sd.RollupCfg, true)
+	l2RPC := env.engine.RPCClient()
+	l2Client, err := host.NewL2Client(l2RPC, env.log, nil, &host.L2ClientConfig{L2ClientConfig: l2ClCfg, L2Head: l2Head})
+	require.NoError(t, err, "failed to create L2 client")
+	l2DebugCl := &host.L2Source{L2Client: l2Client, DebugClient: sources.NewDebugClient(l2RPC.CallContext)}
+	dfault.L2ProcessSource = l2DebugCl
+
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+func Test_ProgramAction_SimpleEmptyChain_HonestClaim_Granite(gt *testing.T) {
+	t := NewDefaultTesting(gt)
+	tp := NewTestParams()
+	dp := NewDeployParams(t, func(dp *e2eutils.DeployParams) {
+		genesisBlock := hexutil.Uint64(0)
+
+		// Enable Cancun on L1 & Granite on L2 at genesis
+		dp.DeployConfig.L1CancunTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisRegolithTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisCanyonTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisDeltaTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisEcotoneTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisFjordTimeOffset = &genesisBlock
+		dp.DeployConfig.L2GenesisGraniteTimeOffset = &genesisBlock
+	})
+	bCfg := NewBatcherCfg()
+	env := NewL2FaultProofEnv(t, tp, dp, bCfg)
+
+	// Build an empty block on L2
+	env.sequencer.ActL2StartBlock(t)
+	env.sequencer.ActL2EndBlock(t)
+
+	// Instruct the batcher to submit the block to L1, and include the transaction.
+	env.batcher.ActSubmitAll(t)
+	env.miner.ActL1StartBlock(12)(t)
+	env.miner.ActL1IncludeTxByHash(env.batcher.LastSubmitted.Hash())(t)
+	env.miner.ActL1EndBlock(t)
+
+	// Finalize the block with the batch on L1.
+	env.miner.ActL1SafeNext(t)
+	env.miner.ActL1FinalizeNext(t)
+
+	// Instruct the sequencer to derive the L2 chain from the data on L1 that the batcher just posted.
+	env.sequencer.ActL1HeadSignal(t)
+	env.sequencer.ActL2PipelineFull(t)
+
+	l1Head := env.miner.l1Chain.CurrentBlock()
+	l2SafeHead := env.engine.l2Chain.CurrentSafeBlock()
+
+	// Ensure there is only 1 block on L1.
+	require.Equal(t, uint64(1), l1Head.Number.Uint64())
+	// Ensure the block is marked as safe before we attempt to fault prove it.
+	require.Equal(t, uint64(1), l2SafeHead.Number.Uint64())
+
+	// Fetch the pre and post output roots for the fault proof.
+	preRoot, err := env.sequencer.RollupClient().OutputAtBlock(context.Background(), l2SafeHead.Number.Uint64()-1)
+	require.NoError(t, err)
+	claimRoot, err := env.sequencer.RollupClient().OutputAtBlock(context.Background(), l2SafeHead.Number.Uint64())
+	require.NoError(t, err)
+
+	// Run the fault proof program from the state transition from L2 block 0 -> 1.
+	programCfg := NewOpProgramCfg(
+		t,
+		env,
+		l1Head.Hash(),
+		preRoot.BlockRef.Hash,
+		common.Hash(preRoot.OutputRoot),
+		common.Hash(claimRoot.OutputRoot),
+		l2SafeHead.Number.Uint64(),
+	)
+	err = host.FaultProofProgram(context.Background(), env.log, programCfg)
+	require.NoError(t, err)
```

Comment:
This is worth introducing a single method for.  I suspect we should be able to just do something a bit like:

```
RunFaultProofProgram(t, env, l2SafeHead.Number.Uint64())
```
and it creates the config and looks up the correct and post output roots for us and use the miner current block as the L1 head. It could then use the options patterns to allow overriding those defaults so you can specify an invalid claim, different L1 head etc. 

I'm not entirely sure where that method should live - I suspect we should create a `actions/proofs` package and it can live there (I'd eventually like to move all the helper/actor code to a `helper` subpackage and only have the actual action tests at the top level).

Reply:
Addressed in #11858